### PR TITLE
[crypto] add a sentinel error for invalid inputs

### PR DESCRIPTION
**Pull request by tarakby**
- add a new specific error type in the crypto package to be returned when inputs are invalid. 
- the new error type is not returned when other unexpected errors occur. 
- use a different return value in the C layer that is returned when an unexpected error occur. This allows the Go layer to return the right error type.

### CODE REVIEW

File: crypto/types.go [hunk 1]
```diff
@@ -100,3 +100,19 @@ const (
 
 // Signature is a generic type, regardless of the signature scheme
 type Signature []byte
+
+// InvalidInputs is an error returned when a crypto API receives invalid inputs.
+// It allows a function caller differentiate unexpected program errors from errors caused by
+// invalid inputs.
+type InvalidInputs struct {
```
Comment (janezpodhostnik) [inline crypto/types.go:107]: Should this type name have a suffix of `Error`?

**Review by janezpodhostnik — APPROVED** [review]
My `c` is rusty, but the code looks ok to me.

The whole review I was asking myself if `InvalidInputs` should rather be defined as a an error that wraps an error. 

Either as 

```go
type InvalidInputs struct {
  Err error
}
```

or as

```go
type InvalidInputs struct {
  Err error
  Message string
}
```
with the Unwrap method defined. That way the errors that caused a InvalidInputs error wouldn't "be lost" (they are not really lost as the message of the InvalidInputs contains the error in string form, but you can't check for a more detailed error by using `error.Is(...)` and `error.As(...)` ). 

I'm referencing cases like this:
```
 newInvalidInputs(fmt.Sprintf("seeding relic failed: %s", err))
```

This would become:
```go
 newInvalidInputs(fmt.Errorf("seeding relic failed: %w", err))
```
or maybe
```go
 newInvalidInputs("seeding relic failed", err)
```

What do you think?

On a similar note using `error.Is(...)` and `error.As(...)` for error comparison/condition would mean tat if the errors were wrapped further the comparison/condition would still work.

File: crypto/types.go (hunk 1 above)
Comment (tarakby) [reply crypto/types.go:107]: I think you're right. The examples I looked at all have the suffix.

**Comment by tarakby** [comment]
Thanks for the review @janezpodhostnik  🙏 

>  I was asking myself if InvalidInputs should rather be defined as a an error that wraps an error.
That's a good suggestion. Looking at the code, I see that `InvalidInputs` error is created most of the times from a simple string, not from an existing error. You can see [this](https://github.com/onflow/flow-go/blob/5541/tarak-review-decode-api/crypto/bls.go#L90) as an example. Your suggestion would work well if `InvalidInputs` was created from an error and wraps it. Does that make sense?

**Comment by janezpodhostnik** [comment]
Sounds good :+1:.

File: crypto/types.go [hunk 2]
```diff
@@ -100,3 +100,19 @@ const (
...
 type Signature []byte
+
+// InvalidInputsError is an error returned when a crypto API receives invalid inputs.
+// It allows a function caller differentiate unexpected program errors from errors caused by
+// invalid inputs.
+type InvalidInputsError struct {
+	message string
+}
+
+// NewAccountAlreadyExistsError constructs a new AccountAlreadyExistsError
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/types.go:111]: oups 🙊

File: crypto/bls.go [hunk 3]
```diff
@@ -182,21 +197,28 @@ func (a *blsBLS12381Algo) decodePrivateKey(privateKeyBytes []byte) (PrivateKey,
...
+		return nil, newInvalidInputsError(fmt.Sprintf(
+			"the input length has to be %d",
+			pubKeyLengthBLSBLS12381))
 	}
 	var pk PubKeyBLSBLS12381
-	if readPointG2(&pk.point, publicKeyBytes) != nil {
-		return nil, errors.New("the input does not encode a BLS12-381 point")
+	err := readPointG2(&pk.point, publicKeyBytes)
+	if err != nil {
+		if _, ok := err.(*InvalidInputsError); ok {
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/bls.go:215]: Nice, I didn't know such a type checker exists! 👌

File: crypto/types.go [hunk 4]
```diff
@@ -100,3 +100,19 @@ const (
...
+func newInvalidInputsError(s string) error {
+	return &InvalidInputsError{message: s}
+}
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/types.go]: Another good suggestion and another thing I didn't know it exists 🙏

**Review by jordanschalm — APPROVED** [review]
🎸

**Pull request merged**